Info improve and add BindConstant

ChrisJefferson opened the pull request:
This patch rearranges how we implement `Info`, to make it more efficient.

Just to give a feeling for the times, I wrote a test which compares an `if` statement against an`Info`.

 `f := function() local i; for i in [1..10000000] do if x = 2 then Print("!!"); fi; od; end;;`

`g := function() local i; for i in [1..10000000] do Info(InfoDebug, 2, "!!"); od; end;;`

`f`: 124ms

`g` in 4.8: 3086ms (24x slower)
`g` in master: 980ms (8x slower)
`g` with this patch: 217ms (1.75x slower)

To go any faster, I think we would have to completely tear out the current system, remove the extensibility, and possibly introduce a `TNUM` for `Info` objects, so we could avoid any error checking.

EDIT: This also includes a new little function, `BindConstant`. This is like `BindGlobal`, but for Constant globals -- it just makes a quick easy way to declare constants. I experiment (I think successfully) with using it to access positional objects efficiently.

stevelinton commented:
Maybe change the title to mention BindConstant. Otherwise looks great to me.
I wrote the original Info code long ago as a placeholder and then forgot about it.

fingolfin commented on src/intrprtr.c:
```diff
@@ -4583,19 +4583,28 @@ void            IntrInfoBegin( void )
 {
-    if (IS_POS_INTOBJ(level) &&
-        CALL_1ARGS(IsInfoClassListRep, selectors) == True) {
+    // Fast-pass the most common failing case
```
I think elsewhere we call(ed) this a "fast-path", not a "fast-pass"

fingolfin commented on src/intrprtr.c:
```diff
@@ -4583,19 +4583,28 @@ void            IntrInfoBegin( void )
-        CALL_1ARGS(IsInfoClassListRep, selectors) == True) {
+    // Fast-pass the most common failing case
+    // The fast-pass never returns True because the GAP function
+    // InfoDecision sets a number of variables as well as returning True
```
This explanation feels a bit backwards to me (but that may just me being dense). I'd suggest something like this: 
> The fast-path only deals with the case were False is returned, as in the
True case, `InfoData.LastClass` and `InfoData.LastLevel` need to be set.


But actually, I think we can get rid of this exception, by getting rid of `InfoData`. In the meantime, though, of course it makes sense to document this exception.

fingolfin commented on lib/info.gi:
```diff
@@ -32,34 +32,36 @@
+INFO_CLASSES := [];
+
+# The most recently called Info statement
+InfoData := rec();
```
I actually think that we can get rid of `InfoData` but that can be done in a later PR (I can take care of it, too).

fingolfin replied:
In the meantime, perhaps say something like this:
> The level and (first) selector of the most recently executed Info statement is stored in InfoData

ChrisJefferson replied:
I will improve the comment. Note that `InfoData` is used by GAPDoc. I tried hard in this commit to not break any way someone might have extended the `Info`, in case anyone has.

ChrisJefferson commented on lib/info.gi:
```diff
@@ -105,42 +107,31 @@ end);
-        pos := Position(InfoData.ClassNames,name);
-        if pos <> fail then
-            return InfoData.InfoClass(pos);
+    atomic readwrite INFO_CLASSES do
```
I thought we were, now that the parser removes them in GAP. However, we should make sure we all agree about that!

fingolfin replied:
Well, to make sure we all agree, "somebody" has to bring it up on e.g. the mailing list...

fingolfin replied:
Ahh, whatever, let's just merge this in. It's deep in internal code, and one can understand the code by just ignoring the `atomic` here. If somebody truly minds, I'll wait for them to bring it up on the mailing list or elsewhere.

ChrisJefferson commented on src/intrprtr.c:
```diff
@@ -4583,19 +4583,28 @@ void            IntrInfoBegin( void )
+        if (IS_INTOBJ(index) && IS_INTOBJ(level)) {
+            Obj res;
+            LT_INTOBJS(res, index, level);
+            if (res == True) {
```
I agree it's complicated, but I couldn't find a shorter way of comparing integers "properly". I could just use the fact that I know that `<` works for a pair of `INTOBJ`s.

ChrisJefferson commented on lib/info.gi:
```diff
@@ -32,34 +32,36 @@
 ##
-##  these are all lists, and the first two should have the same length, 
-##  which defines the number of Info classes that exist
+##  ClassNum is placed at the end so the length of the list never changes
```
I'll explain it better. I don't want to put optional members at the end, because then the list of the length can change when those optional members become assigned, which doesn't work for IsAtomicPositionalObjects, which use fixed length atomic lists (that comment really is too short).

ChrisJefferson commented on src/intrprtr.c:
```diff
@@ -4583,19 +4583,28 @@ void            IntrInfoBegin( void )
-    }
-    else {
-        return CALL_2ARGS(InfoDecision, selectors, level);
+#endif
```
Because `ELM_LIST` doesn't work on positional objects, or atomic positional objects.

If we make it work, then given a positional object `X[3]` starts working at the GAP level, rather than the `X![3]` we expect people to use.

However, some kind of unified access would be nice, as this code is fairly horrible and I wouldn't want to start writing this regularily.

fingolfin replied:
OK, that makes sense.

ChrisJefferson commented:
This is now improved, fixed all comments, and added tests, except for the slightly horrible "ifdef HPCGAP" checking inside lists bit, where we should indeed come up with something better (or I should find out how I'm supposed to read members of positional objects)

fingolfin commented on lib/global.g:
```diff
@@ -145,6 +145,22 @@ BIND_GLOBAL := function( name, val)
+    MAKE_CONSTANT_GLOBAL(name);
+    return val;
+end;
+MAKE_READ_ONLY_GLOBAL("BIND_CONSTANT");
```
Nitpick: Why not use `BIND_GLOBAL` for defining `BIND_CONSTANT`, i.e. `BIND_GLOBAL("BIND_CONSTANT", ...)` ?

fingolfin commented on lib/global.gd:
```diff
@@ -196,18 +196,21 @@ DeclareGlobalFunction("MakeConstantGlobal");
-##  <A>val</A>, provided it is writable, and makes it read-only.
+##  <Ref Func="BindGlobal"/> and <Ref Func="BindConstant"/> set the global
+##  variable named by the string <A>name</A> to the value
+##  <A>val</A>, provided it is writable. <Ref Func="BindGlobal"/> makes
```
The "it" in "it is writable" is ambiguous. Perhaps "provided that variable is writable", or even "provided that variables is unbound, or bound and writable".

[Aha, but I just realize the ambiguous text was there before. *sigh*]

fingolfin commented on src/intrprtr.c:
```diff
@@ -4583,19 +4583,28 @@ void            IntrInfoBegin( void )
-    else {
-        return CALL_2ARGS(InfoDecision, selectors, level);
+#endif
+        if (IS_INTOBJ(index) && IS_INTOBJ(level)) {
```
You could use `ARE_INTOBJS` here. But the above is of course fine, too.

ChrisJefferson commented:
This should wait until after #1789

stevelinton commented:
Looks like this needs a rebase and a trivial conflict resolved. Then I'd be happy to merge it -- trying to get as many performance patches in as is safe.

fingolfin commented on lib/global.gd:
```diff
@@ -196,18 +196,21 @@ DeclareGlobalFunction("MakeConstantGlobal");
+##  <Ref Func="BindGlobal"/> and <Ref Func="BindConstant"/> set the global
+##  variable named by the string <A>name</A> to the value
+##  <A>val</A>, provided it is writable. <Ref Func="BindGlobal"/> makes
+##  the resulting variable read-only, which <Ref Func="BindConstant"/> makes
```
Just noticed: "which" -> "while"

stevelinton commented:
I'd be happy to see this rebased and merged. It's another useful performance improvement and makes code overall much nicer.

fingolfin commented:
Can be merged from my POV once tests passed

ChrisJefferson commented:
Now passing, and the bindconstant removed.
